File: src/sort/insertionSort.cc
#include <vector>
#include <iostream>

using std::vector;

class Solution {
public:
    void insertionSort(vector<int>& nums) {
      for (std::size_t i = 1; i < nums.size(); i++) {
        int value = nums[i];
        auto j = i - 1;
        while (j>=0 && nums[j] > value) {
          nums[j+1] = nums[j];
          j--;
        }

        nums[j+1] = value;
      }
    }
};

int main() {
  vector<int> nums = {5, 2, 4, 6, 1, 3};
  Solution().insertionSort(nums);
  for (auto i : nums) {
    std::cout << i << " ";
  }
  std::cout << std::endl;
  return 0;

}

File: src/sort/go/insertion_sort.go
package main

// func main() {
// 	nums := []int{6, 4, 2, 5, 3, 1}
// 	fmt.Println(nums)
// 	res := insertion_sort(nums)
// 	fmt.Println(res)
// }

func insertion_sort(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}

	for i := 1; i < len(nums); i++ {
		value := nums[i]
		var j int = i - 1

		//查到插入位置
		for ; j >= 0; j-- {
			if value < nums[j] {
				nums[j+1] = nums[j]
			} else {
				break
			}
		}

		//插入对应位置
		nums[j+1] = value
	}

	return nums
}
